Various matrix generator improvements

neil-lindquist opened the pull request:
This is a miscellaneous set of improvements coming from the BEAM work.
* Fix some invalid `omp parallel for` usage in the `diag`, `svd`, `heev`, and `poev` generators
* Fix an indexing issue in the `ris` generator
* Add some extra matrix generators (`chebspec` and `kms` from Matlab gallery, `ZielkeNS` from Higham's new AnyMatrix collection, `randr` as a random matrix from the Rademacher distribution)
* Add some seed preprocessing to prevent similar seeds from having similar A[1,1] entries
* Add support for Options
  * Key use is doing `geqrf` and `unmqr` in the `svd` and `heev` generators on device
  * Currently used by only the `gesv` tester when `Target::Devices`
* Add missing `A.tileGetForWriting` calls to structured matrix generators

Relatedly, there's the work in progress [PR#97](https://bitbucket.org/icl/slate-dev/pull-requests/97) which was to add a way to test different matrices with `--repeat` in a reproducible way.  (The current default seed of `-1` gives a different matrix each time, but is not reproducible.)
If that's still wanted, I could get it finalized and added to this PR.

mgates3 commented on test/test_gesv.cc:
```diff
@@ -221,8 +221,9 @@ void test_gesv_work(Params& params, bool run)
-    slate::generate_matrix(params.matrixB, B);
+    slate::Options matgen_opts = {{slate::Option::Target, target}};
+    slate::generate_matrix(params.matrix,  A, matgen_opts);
+    slate::generate_matrix(params.matrixB, B, matgen_opts);
```
Why not just use the `opts` used for `getrf`, etc.?

neil-lindquist replied:
I was trying to get identical matrices, and `panel_threads` and `ib` both modify the order of operations (and thus round off).  Although, `geqrf` is already slightly non-deterministic for identical options, so it's probably a moot point.

I can switch it to `opts` if you think that's better.

mgates3 replied:
Makes sense to keep identical matrices.

neil-lindquist commented:
I discovered that the structured matrix generators were missing `tileGetForWriting`'s.  So, I'm adding that as well.

neil-lindquist commented:
I updated this to fix the merge conflicts from merging the Philox PR.

I also removed a commit and added a commit.
* Philox removed the need to "Add some seed preprocessing to prevent similar seeds from having similar A[1,1] entries", so that's been dropped
* Minor cleanup to the file
  * Removed a redundant declaration for `split`
  * Improve the clarity of a comment
  * Improved variable name consistency
  * Use an explicit pi constant instead of a computed value
